### PR TITLE
Fix #1869: Allow Scala.js-defined JS objects.

### LINKED ISSUE #1869: Allow Scala.js-defined JS objects
There is no reason whatsoever not to allow them, so they should be allowed. The semantics are obvious: it's a lazy val whose value is an instance of a Scala.js-defined JS module class.

### PR DESCRIPTION
**Pull request by sjrd**

### CODE REVIEW

**Comment by sjrd** [comment]
Review by @gzm0 if you have time tonight or tomorrow; otherwise @nicolasstucki. We're trying to schedule 0.6.5 with the rest of Scala.js-defined JS stuff for this Friday/week-end, and since this PR touches many parts of the code, I'll have to base the other ones on top of this one, I fear; so it would be good for this to get in relatively quickly.

The other PRs will be much simpler.

**Comment by sjrd** [comment]
Actually even this PR is *mostly* tests and one-line changes.

File: compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala [hunk 1]
```diff
@@ -338,6 +339,16 @@ abstract class PrepJSInterop extends plugins.PluginComponent
...
         }
+
+        if (sym.isModule && (enclosingOwner is OwnerKind.JSNonNative)) {
+          def shouldBeExposed: Boolean = {
+            !sym.isSynthetic &&
+            !isPrivateMaybeWithin(sym)
+          }
+          if (shouldBeExposed)
+            sym.addAnnotation(ExposedJSMemberAnnot)
+        }
```
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala]: This part is to expose `objects` *inside* Scala.js-defined JS classes/objects.

**Comment by gzm0** [comment]
Reviewing this now

File: compiler/src/test/scala/org/scalajs/core/compiler/test/ScalaJSDefinedTest.scala [hunk 2]
```diff
@@ -400,6 +472,26 @@ class ScalaJSDefinedTest extends DirectTest with TestHelpers {
...
+      def foo: Int = js.native
+      @JSName(A.b)
+      def bar: Int = js.native
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:12: error: The argument to JSName must be a literal string
+      |      @JSName(A.a)
+      |       ^
+    """
```
Comment (gzm0) [inline compiler/src/test/scala/org/scalajs/core/compiler/test/ScalaJSDefinedTest.scala]: These test cases are potentially unstable, since there are other errors in the class... (I don't know how the one for the class slipped my review)

File: ir/src/main/scala/org/scalajs/core/ir/Trees.scala [hunk 3]
```diff
@@ -589,6 +589,12 @@ object Trees {
     val tpe = AnyType
   }
 
+  /** Like [[LoadModule]] but for a JS module class. */
+  case class JSLoadModule(cls: ClassType)(
+      implicit val pos: Position) extends Tree {
+    val tpe = AnyType
+  }
+
```
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/core/ir/Trees.scala]: Consider calling this `LoadJSModule`. It loads a JS module. It is not a JavaScript way of loading a module (such a thing does not really exist, does it...)

File: tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala [hunk 4]
```diff
@@ -69,10 +69,12 @@ final class Refiner(semantics: Semantics, outputMode: OutputMode) {
...
     val kind = {
-      if (classDef.kind == ClassKind.ModuleClass && !info.isModuleAccessed)
-        ClassKind.Class
-      else
+      if (classDef.kind.hasModuleAccessor && !info.isModuleAccessed) {
+        if (classDef.kind == ClassKind.ModuleClass) ClassKind.Class
+        else ClassKind.JSClass
+      } else {
         classDef.kind
+      }
```
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala]: Have you considered unifying this code with the one in the linker? It seems like a good candidate for a helper.

File: compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala (hunk 1 above)
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala]: Isn't this change unrelated to Scala.js defined objects? This will also expose normal objects (which is correct). I think this should be in a separate commit.

**Comment by gzm0** [comment]
Please add tests for exposing of objects inside Scala.js defined classes. (Preferably in a separate commit).

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala [hunk 5]
```diff
@@ -458,12 +458,24 @@ abstract class GenJSCode extends plugins.PluginComponent
 
       gen(cd.impl)
 
+      // Generate exported module accessors
+      val exports = {
+        // Generate exported constructors or accessors
+        val exportedAccessors =
+          if (isStaticModule(sym)) genModuleAccessorExports(sym)
+          else Nil
```
Comment (gzm0) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: I'm assuming this prepares for exporting of Scala.js defined *classes*.

**Comment by gzm0** [comment]
That's all. Neat how little change this needs :)

File: compiler/src/test/scala/org/scalajs/core/compiler/test/ScalaJSDefinedTest.scala (hunk 2 above)
Comment (sjrd) [inline compiler/src/test/scala/org/scalajs/core/compiler/test/ScalaJSDefinedTest.scala]: Ouch, indeed. Copy paste error in the first place, I guess.

File: ir/src/main/scala/org/scalajs/core/ir/Trees.scala (hunk 3 above)
Comment (sjrd) [inline ir/src/main/scala/org/scalajs/core/ir/Trees.scala]: Should I then also rename `JSLoadConstructor` -> `LoadJSConstructor`? The same reasoning applies to both.

File: compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala (hunk 1 above)
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/PrepJSInterop.scala]: Yes, you are right. I will make a separate commit with its own tests.

File: tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala (hunk 4 above)
Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala]: If you have a suggestion about *where* to put such a helper? `ClassKind`'s companion object?

Eventually (probably in the next major version), I want to entirely refactor `ClassKind`s anyway. They now conflate several axes that have become orthogonal. For example, I would reify `hasModuleAccessor` as a separate flag.

File: compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala (hunk 5 above)
Comment (sjrd) [inline compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala]: Yes.

**Comment by sjrd** [comment]
Addressed the 2 commits stuff & the "unstable" test.
Waiting for confirmation on the two other items before changing stuff (`LoadJSConstructor`/`LoadJSModule` and the `removeModuleAccessor` helper).

File: ir/src/main/scala/org/scalajs/core/ir/Trees.scala (hunk 3 above)
Comment (gzm0) [inline ir/src/main/scala/org/scalajs/core/ir/Trees.scala]: Indeed. It didn't bother me that much, since "loading" a constructor is nonsensical in non-js land anyways... But maybe it is more consistent indeed.

File: tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala (hunk 4 above)
Comment (gzm0) [inline tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala]: > ClassKind's companion object?

Yeah... I guess that's best. It is a step in the right direction probably. You probably want to keep it private to `optimizer` though...

Btw. Is it really a good idea to have a `hasModuleAccessor` flag? This will lead to combinations which are not allowed... (anyways, not subject to discussion right now).

Comment (sjrd) [inline tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/Refiner.scala]: > You probably want to keep it private to optimizer though...

`ClassKind` is in `ir/` ... Can't do.

**Comment by sjrd** [comment]
Updated. All concerns should be addressed, now, unless I missed something.

**Comment by sjrd** [comment]
retest this please

**Comment by gzm0** [comment]
LGTM
